### PR TITLE
TimePicker: Close overlay content when toolbar button is clicked for the second time

### PR DESCRIPTION
**Pull request by juanicabanas**
**What is this feature?**

It closes the overlay content when the toolbar button is clicked for the second time

**Why do we need this feature?**

This is a bug fix, since a double set state is being triggered and, therefore, the overlay content keeps open.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

| Previous behavior  | Fixed behavior |
| ------------- | ------------- |
| <video src="https://github.com/grafana/grafana/assets/11707172/4b2df5cd-1e3c-4930-8fe3-b4a1e6edefc8"> | <video src="https://github.com/grafana/grafana/assets/11707172/1a97ed6b-b733-4896-8501-746eb27a9f9f">  |

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.test.tsx [hunk 1]
```diff
@@ -1,10 +1,13 @@
-import { render } from '@testing-library/react';
+import { fireEvent, render, screen } from '@testing-library/react';
```
Comment (ashharrison90) [inline packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.test.tsx:1]: can we use `userEvent` here instead of `fireEvent` 🙏

**Review by ashharrison90 — CHANGES_REQUESTED** [review]
hi @juanicabanas 👋 

thanks for spotting this!

i would prefer if we didn't move the `overlayProps` since i think it'll have some a11y implications... and the button isn't really part of the overlay, so it doesn't make conceptual sense either.

instead, let's utilise `useOverlay`'s `shouldCloseOnInteractOutside`, and do something like:

```tsx
  const buttonRef = useRef<HTMLButtonElement>(null);
  const { overlayProps, underlayProps } = useOverlay({
    onClose,
    isDismissable: true,
    isOpen,
    shouldCloseOnInteractOutside: (element) => {
      return !buttonRef.current?.contains(element);
    },
  }, ref);
```
and attach `buttonRef` to the `Tooltip` wrapping the button (this is important! you have to pass it to the wrapping tooltip else it won't get forwarded to the button correctly)

at this point it probably makes sense to rename the existing `ref` to `overlayRef` as well

shout if anything doesn't make sense 👍

**Comment by juanicabanas** [comment]
> hi @juanicabanas 👋
> 
> thanks for spotting this!
> 
> i would prefer if we didn't move the `overlayProps` since i think it'll have some a11y implications... and the button isn't really part of the overlay, so it doesn't make conceptual sense either.
> 
> instead, let's utilise `useOverlay`'s `shouldCloseOnInteractOutside`, and do something like:
> 
> ```tsx
>   const buttonRef = useRef<HTMLButtonElement>(null);
>   const { overlayProps, underlayProps } = useOverlay({
>     onClose,
>     isDismissable: true,
>     isOpen,
>     shouldCloseOnInteractOutside: (element) => {
>       return !buttonRef.current?.contains(element);
>     },
>   }, ref);
> ```
> 
> and attach `buttonRef` to the `Tooltip` wrapping the button (this is important! you have to pass it to the wrapping tooltip else it won't get forwarded to the button correctly)
> 
> at this point it probably makes sense to rename the existing `ref` to `overlayRef` as well
> 
> shout if anything doesn't make sense 👍

@ashharrison90  yeah, makes a lot of sense. didn't know about `shouldCloseOnInteractOutside` . I'll make those changes

File: public/app/core/components/TimePicker/TimePickerWithHistory.test.tsx [hunk 2]
```diff
@@ -116,11 +116,11 @@ describe('TimePickerWithHistory', () => {
     await userEvent.click(screen.getByLabelText(/Time range selected/));
 
     for (const [inputFrom, inputTo] of inputRanges) {
-      await userEvent.click(screen.getByLabelText(/Time range selected/));
```
Comment (ashharrison90) [inline public/app/core/components/TimePicker/TimePickerWithHistory.test.tsx:119]: minor nit: think it makes more sense to leave this where it is and remove the click outside the loop on line 116 instead? 🤔

**Review by ashharrison90 — APPROVED** [review]
couple of minor nits, otherwise lgtm 👍

**Pull request merged**